### ISSUE TITLE
Pre-validate uniqueness.

### ISSUE THREAD

**Issue by tomchristie** [encode/orm#11]
Eg. If `username` is a unique field, then `User.objects.create(username=...)` should ideally:

* Run inside a transaction.
* Run a query to check no existing username already exists. Raise an exception otherwise.
* Perform the insert.

This is far more helpful than raising an integrity error.

**Comment by florimondmanca** [comment]
I'm interested in working on this!

Which kind of exception should be raised? `ValidationError`? A custom `IntegrityError`?

**Comment by tomchristie** [comment]
A `typesystem.ValidationError` yup. At least for now.

> Also, could you explain why the transaction is required?

Let's not worry about that at the moment. But yeah, ideally you'd want that, otherwise there's a race condition between "We've checked uniqueness" and "we've performed the insert". If you get caught by that race, then you can end up having an DB error raised, rather than a nice validation error.

**Comment by tomchristie** [comment]
I mean Django doesn't cover it I think, nor REST framework, so it's not really a big deal, and the *integrity* of the data is still safe, so it's really not a big deal in practice.

**Comment by florimondmanca** [comment]
Yes, I see. I also suppose we're safe since the data integrity is ensured by the database anyway (at least for MySQL and PostgreSQL).
